Add question bank filter API guide

marxjohnson opened the pull request:
This functionality was added in MDL-72321. MDL-78220 will add several new filters, so this guide may expand as the children of that tracker are developed.

timhunt reviewed (approved):
Thanks Mark. This looks like a good start to documenting this.

timhunt commented:
Ah! but I see taht the automated checks need to be appeased.

andrewnicols commented on docs/apis/plugintypes/qbank/filters.md:
````diff
@@ -0,0 +1,215 @@
+
+Create a `condition` class within your plugin's namespace. For a plugin called `qbank_myplugin` this would look something like:
+
+```php
````
I find it's really handy to specify the location of the class in the title, for example:
```suggestion
```php title="question/bank/myplugin/classes/myfilter_condition.php"
```

marxjohnson replied:
Done

andrewnicols commented on docs/apis/plugintypes/qbank/filters.md:
````diff
@@ -0,0 +1,215 @@
+
+Define the `get_name` method, which returns the label displayed in the filter UI.
+
+```php
+    public function get_name(): string {
+        return get_string('myfilter_name', 'myplugin');
+    }
````
Since this example is in isolation, you could either:
* add the class wrapper to make it clearer; or
* remove the indent to make it look cleaner in the built code.

```suggestion
```php
class myfilter_condition extends condition {
    public function get_name(): string {
        return get_string('myfilter_name', 'myplugin');
    }
}
```


Or

```suggestion
```php
public function get_name(): string {
    return get_string('myfilter_name', 'myplugin');
}
```

andrewnicols replied:
You may also want to use the title attribute here, but that's up to you. If you do I wouldn't set it to the filename, but a short sumarry of what you're doing, for example:

```
title="Define the filter name"

andrewnicols commented on docs/apis/plugintypes/qbank/filters.md:
````diff
@@ -0,0 +1,215 @@
+}
+```
+
+Define the `get_name` method, which returns the label displayed in the filter UI.
````
You may also want to consider adding parens to method names:
```suggestion
Define the `get_name()` method, which returns the label displayed in the filter UI.
```

Your call but if you do please do so for all of them.

andrewnicols commented on docs/apis/plugintypes/qbank/filters.md:
```diff
@@ -0,0 +1,215 @@
+In your condition class, define `get_filteroptions()` which returns an object containing the current filter options. You will
+probably want to add some code to the constructor to read in the current filter options, and some code the `build_query_from_filter`
+to use the option. See
+[`qbank_managecategories\category_condition`](https://github.com/moodle/moodle/blob/main/question/bank/managecategories/classes/category_condition.php#L331)
```
Consider making this link point to a specific commit hash, or to just the file as a whole?

andrewnicols reviewed (changes requested):
These docs look generally great. I've left a few comments -- see what you think.

Since these are docs for a 4.3 feature we should backport to 4.3 and 4.4 using the `versioned_docs` folders. If you need help with this I can do this on the PR -- just let me know.

Thanks

marxjohnson commented:
@andrewnicols Thanks for your comments, I have made those changes and re-worked the ordering of things a bit to make it clearer which bits are compulsory, and which are more advanced. I've included those changes as fixups so I can squash them if you're happy.

I've also added an extra commit to cover the validation support I've added in MDL-83862.

marxjohnson commented:
@andrewnicols Could you do that `versioned_docs` thing you mentioned? I've not done that before I'd be grateful of the help!

marxjohnson commented:
Now this has been merged, should it be showing up on the live site or is there there a "release" that we need to wait for? I don't see it here: https://moodledev.io/docs/5.0/apis/plugintypes/qbank
